src/backend/metal/src/window.rs

```diff
@@ -355,13 +375,97 @@ impl hal::Surface<Backend> for Surface {
+    unsafe fn configure_swapchain(
+        &mut self, device: &Device, config: SurfaceSwapchainConfig
+    ) -> Result<(), CreationError> {
+        info!("configure_swapchain {:?}", config);
+
+        let caps = &device.shared.private_caps;
+        let mtl_format = caps
+            .map_format(config.format)
+            .expect("unsupported backbuffer format");
+        self.swapchain_format = mtl_format;
+
+        let render_layer_borrow = self.inner.render_layer.lock();
+        let render_layer = *render_layer_borrow;
+        let framebuffer_only = true;
+        let display_sync = config.present_mode != PresentMode::Immediate;
+        let is_mac = caps.os_is_mac;
+        let can_set_next_drawable_timeout = if is_mac {
+            caps.has_version_at_least(10, 13)
+        } else {
+            caps.has_version_at_least(11, 0)
+        };
+        let can_set_display_sync = is_mac && caps.has_version_at_least(10, 13);
+        let drawable_size = {
+            let bounds: CGRect = msg_send![render_layer, bounds];
+            bounds.size
+        };
+
+        let device_raw = device.shared.device.lock().as_ptr();
+        msg_send![render_layer, setDevice: device_raw];
+        msg_send![render_layer, setPixelFormat: mtl_format];
+        msg_send![render_layer, setFramebufferOnly: framebuffer_only];
+
+        // this gets ignored on iOS for certain OS/device combinations (iphone5s iOS 10.3)
+        msg_send![render_layer, setMaximumDrawableCount: config.image_count as u64];
+
+        msg_send![render_layer, setDrawableSize: drawable_size];
+        if can_set_next_drawable_timeout {
+            msg_send![render_layer, setAllowsNextDrawableTimeout:false];
+        }
+        if can_set_display_sync {
+            msg_send![render_layer, setDisplaySyncEnabled: display_sync];
+        }
+
+        Ok(())
+    }
+
+    /// Acquire a new swapchain image for rendering.
+    ///
+    /// May fail according to one of the reasons indicated in `AcquireError` enum.
+    ///
+    /// # Synchronization
+    ///
+    /// The acquired image is available to render. No synchronization is required.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    ///
+    /// ```
+    unsafe fn acquire_image(
+        &mut self,
+        _timeout_ns: u64, //TODO: use the timeout
+    ) -> Result<(Self::SwapchainImage, hal::SwapchainImageId, Option<Suboptimal>), hal::AcquireError> {
+        // for the drawable & texture
+        let render_layer_borrow = self.inner.render_layer.lock();
+        let (drawable, raw) = autoreleasepool(|| {
+            let drawable: &metal::DrawableRef = msg_send![*render_layer_borrow, nextDrawable];
+            assert!(!drawable.as_ptr().is_null());
+            let texture: &metal::TextureRef = msg_send![drawable, texture];
+            (drawable.to_owned(), texture.to_owned())
+        });
+
+        let id = raw.as_ptr() as usize as hal::SwapchainImageId; //HACK
```

Comment:
Would we be able to use [`drawableID`](https://developer.apple.com/documentation/metal/mtldrawable/2806860-drawableid?language=objc) here?

Reply:
According to the docs, this would just be monotonically increasing. Is there use for this ID then? The original idea behind the ID was that the users can keep associated framebuffers... I'd really like to get rid of this ID semantics though, it's not very sound. The main blocker is a potential use of the swapchain image in a framebuffer that has other images :/

Reply:
I interpreted the documentation as the ID increasing every time a new drawable is created, but if an existing drawable is recycled, no new drawable would be created, and thus the ID would be the same. I notice that it doesn't explicitly state this, so I would have to test what it returns. Regardless, I realize now that `drawableID` is only available on iOS 10.3+ and tvOS 10.3+. It is not available on macOS.

Reply:
It seems borderline impossible to implement unique id without running into ABA. The metal backend would still have to track MTLTextures somehow.